src/audio/dai.c

```diff
@@ -64,6 +74,47 @@ struct dai_data {
 	uint64_t wallclock;	/* wall clock at stream start */
 };
 
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data);
+
+static struct dai_group *dai_add_group(struct comp_dev *dev, uint8_t group_id)
+{
+	struct list_item *dai_groups = &sof_get()->dai_groups;
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct list_item *group_item;
+	struct dai_group *group;
+
+	if (dd->group) {
+		comp_info(dev, "dai_add_group(), DAI already in group");
+		return dd->group;
+	}
+
+	list_for_item(group_item, dai_groups) {
+		group = container_of(group_item, struct dai_group, list);
+		if (group && group->group_id == group_id)
+			break;
+		else
+			group = NULL;
+	}
+
+	if (!group) {
+		group = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
+				sizeof(*group));
+
+		list_item_prepend(&group->list, dai_groups);
+		group->group_id = group_id;
+		group->trigger_cmd = -1;
+	}
+
+	group->num_dais++;
```

Comment:
is there any risk of concurrency that might impact the list insertion and num_dais increment? should this be protected with a lock?

Reply:
Is topology loading concurrent? Do we _load new components_ concurrently? Or component loading is already serialized, even if it's in parallel with active playback/capture?

Reply:
I'm not sure I'm afraid, good question.

Reply:
I'm a bit confused by a thing: You claimed groups are identified by group ID and by core, yet you allocate each group ID only once (and the structure for all groups with the same ID is shared). Shouldn't you have separate structures with separate counters? Or you can just... assume a group ID maps to a single core number?

Reply:
Blocking on this inconsistency until explained, sorry.

Reply:
sof_get() returns a single global sof structure. You access a single shared groups list. You search in this shared list and only check group ID. I don't see where the separation between cores comes in, does it come in kernel side? Does the kernel itself ensure that you cannot get a group ID matching for DAIs on multiple cores?

Reply:
Ah, you're right. I actually thought the structure was separate for each core, but it's not. Alright then